Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,176 @@
+
+*As proposed there would be no support for unloading components. For discussion on possible solutions see open issues below.*
+
+### Locate `hostfxr`
+``` C++
+typedef void(* nethost_get_hosfxr_path_result_fn)(const char * hostfxr_path);
+
+int nethost_get_hostfxr_path(
+        nethost_get_hostfxr_path_result_fn result,
+        const char * assembly_path); // Optional
+```
+
+This API locates the `hostfxr` and returns its path by calling the `result` function. (This callback design is chosen so that it's clear and easy to define memory ownership.)
+
+`assembly_path` is optional:
+* If not specified, the "muxer" (`dotnet.exe`) behavior is used to locate the `hostfxr`
+* If specified, the `apphost` behavior is used
+
+
+## Improve API to run application
+New API will be added to `hostfxr`.
+
+``` C++
+struct hostfxr_runtime_properties
+{
+    int count;
+    const char ** keys;
+    const char ** values;
+};
+
+struct hostfxr_parameters
+{
+    int size;
+    const char * host_path;
+    const char * dotnet_root;
+    const char * app_path;
+    hostfxr_runtime_properties additional_properties;
+};
+
+int hostfxr_main_with_parameters(
+        const int argc,
+        const char* argv[],
+        const hostfxr_parameters * parameters);
+```
+
+This new API supercedes the existing `hostfxr_main` and `hostfxr_main_startupinfo`. The function is the main entry point for running applications and dotnet commands. It encapsulates both the `muxer` functionality for running apps via `dotnet app.dll` or `dotnet exec app.dll`, as well as running SDK commands like `dotnet new` and so on.
+
+The proposed behavior is the same as for the existing entry points `hostfxr_main` and `hostfxr_main_startupinfo` with the addition of `parameters` structure. This is meant to:
+* Add the ability to specify additional runtime properties which will be added to the property bag pass to the runtime during its initialization.
+* Make the API extensible without adding new entry point. In the future the parameters structure can be extended with additional members.
+
+The ability to specify additional runtime properties introduces potential collisions with runtime properties initialized by the hosting layer. This ability is not new, it's already possible to specify arbitrary set of additional runtime properties through the `configProperties` section of the `.runtimeconfig.json`. The existing behavior for collisions is somewhat undefined by it effectively means:
+* Properties supplied by the host are added first
+* Any additional properties are added after these - duplicates are retained
+* Runtime initialization will use the last value of any given property it finds for properties it needs to start running managed code
+* Later on during runtime initialization duplicates will cause a failure and the runtime will fail to start with `0x80070057 (E_INVALIDARG)` (internally the failure is due to adding the same key twice to a `Dictionary<K,V>`).
```

Comment:
While dict.Add(key, value) throws, `dict[key] = value;` doesn't throw on overwrite.
Is there any particualr reason you are advocating leaving the current behavior untouched?  
Or are you describing what will happen with old runtimes?

Reply:
I think it's OK to leave it as is. The fact that it always fails means that one day if we come up with a solution we can introduce a change which will make it not fail - without it being a breaking change.
As noted in the discussion above, the problem is relatively tricky and so I want to be able to stage this if needed.